app/services/service_provider_request_proxy.rb

```diff
@@ -0,0 +1,80 @@
+# Drop in replacement for ServiceProviderRequest. Moves us from Postgres to Redis
+# To manage the migration and still respect in flight transactions code will default
+# to checking the db if no redis object is available. Following release can remove db dependence.
+# The .last uuid written is stored only in test mode to support existing specs
+# To migrate code simply replace ServiceProviderRequest with ServiceProviderRequestProxy
+class ServiceProviderRequestProxy
+  REDIS_KEY_PREFIX = 'spr:'.freeze
+  REDIS_LAST_UUID_KEY = 'spr_last_uuid'.freeze
+  DEFAULT_TTL_HOURS = 24
+
+  def self.from_uuid(uuid)
+    find_by(uuid: uuid) || NullServiceProviderRequest.new
+  rescue ArgumentError # a null byte in the uuid will raise this
+    NullServiceProviderRequest.new
+  end
+
+  def self.delete(guid)
+    cache.delete(key(guid)) if guid
+  end
+
+  def self.find_by(uuid:)
+    return unless uuid
+    obj = cache.read(key(uuid))
+    return hash_to_spr(obj, uuid) if obj
+    ServiceProviderRequest.find_by(uuid: uuid)
+  end
+
+  def self.find_or_create_by(uuid:)
+    obj = find_by(uuid: uuid)
+    return obj if obj
+    spr = ServiceProviderRequest.new(uuid: uuid)
+    yield(spr)
+    create(uuid: uuid,
+           issuer: spr.issuer,
+           url: spr.url,
+           loa: spr.loa,
+           requested_attributes: spr.requested_attributes)
+  end
+
+  def self.create(hash)
+    uuid = hash[:uuid]
+    obj = { issuer: hash[:issuer],
+            url: hash[:url],
+            loa: hash[:loa],
+            requested_attributes: hash[:requested_attributes] }
+    cache.write(key(uuid), obj)
+    cache.write(REDIS_LAST_UUID_KEY, uuid) if Rails.env.test?
+    hash_to_spr(obj, uuid)
+  end
+
+  def self.create!(hash)
+    self.create(hash)
+  end
+
+  def self.last
+    uuid = cache.read(REDIS_LAST_UUID_KEY)
```

Comment:
what do you think about just keeping an in-memory link to the "last object"? Like a class property? Is there a particular part of the persistence that you care about?

Reply:
Related: Do we need to clear the `.last` before/after each relevant test? What do you expect the behavior will be for this `.last` once we move to redis?

Reply:
Once we are on Redis `.last` will not be needed and the specs will need something like `last_uuid` which can replace `.last`